File: Grind169/Week2/169-Majority-Element/solution.cpp
#include <vector>

using namespace std;

class Solution
{
public:
    int majorityElement(vector<int> &nums)
    {
        int cnt = 1;
        int res = nums[0];
        int major = nums.size() / 2;

        for (int i = 1; i < nums.size(); i++)
        {
            if (nums[i] == res)
            {
                cnt += 1;
            }
            else
            {
                cnt -= 1;
            }

            if (cnt == 0)
            {
                res = nums[i];
                cnt = 1;
            }

            if (cnt > major)
            {
                return res;
            }
        }

        return res;
    }
};

File: Grind169/Week2/169-Majority-Element/solution.py
from typing import List


class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        # Use Moore's Voting Algorithm
        # https://ithelp.ithome.com.tw/articles/10213285
        cnt = 1
        res = nums[0]

        for n in nums[1:]:
            if n == res:
                cnt += 1
            else:
                cnt -= 1

            if cnt == 0:
                res = n
                cnt = 1

        return res
